components/logins/src/db.rs

```diff
@@ -301,9 +302,54 @@ impl LoginDb {
         rows.collect::<Result<_>>()
     }
 
-    pub fn get_by_hostname(&self, hostname: &str) -> Result<Vec<Login>> {
-        let mut stmt = self.db.prepare_cached(&GET_ALL_BY_HOSTNAME_SQL)?;
-        let rows = stmt.query_and_then(&[hostname], Login::from_row)?;
+    pub fn get_by_base_domain(&self, base_domain: &str) -> Result<Vec<Login>> {
+        // We first parse the input string as a host so it is normalized.
+        let base_host = match Host::parse(base_domain) {
+            Ok(d) => d,
+            Err(e) => {
+                // don't log the input string as it's PII.
+                log::warn!("get_by_base_domain was passed an invalid domain: {}", e);
+                return Ok(vec![]);
+            }
+        };
+        // We just do a linear scan. Another option is to have an indexed
+        // reverse-host column or similar, but current thinking is that it's
+        // extra complexity for (probably) zero actual benefit given the record
+        // counts are expected to be so low.
+        // A regex would probably make this simpler, but we don't want to drag
+        // in a regex lib just for this.
+        let mut stmt = self.db.prepare_cached(&GET_ALL_SQL)?;
+        let rows = stmt
+            .query_and_then(NO_PARAMS, Login::from_row)?
+            .filter(|r| {
+                let login = r
+                    .as_ref()
+                    .ok()
+                    .and_then(|login| Url::parse(&login.hostname).ok());
+                let this_host = login.as_ref().and_then(|url| url.host());
+                match (&base_host, this_host) {
+                    (Host::Domain(base), Some(Host::Domain(look))) => {
+                        // a fairly long-winded way of saying
+                        // `login.hostname == base_domain ||
+                        //  login.hostname.ends_with('.' + base_domain);`
+                        let mut rev_input = base.chars().rev();
+                        let mut rev_host = look.chars().rev();
+                        loop {
+                            match (rev_input.next(), rev_host.next()) {
+                                (Some(ref a), Some(ref b)) if a == b => continue,
+                                (None, None) => return true, // exactly equal
+                                (None, Some(ref h)) => return *h == '.',
+                                _ => return false,
+                            }
```

Comment:
Neat. IIUC this doesn't actually materialize the reversed strings either, right, it should just construct an iterator that walks them together in reverse?

Reply:
yeah, IIUC that's correct. In many ways it's much easier using languages where a little extra cost isn't frowned on - no one in the python/js universes would blink at the "obvious" way to write that :)